pkg/awsutils/awsutils.go

```diff
@@ -1515,3 +1520,22 @@ func (cache *EC2InstanceMetadataCache) IsUnmanagedENI(eniID string) bool {
 	}
 	return false
 }
+
+func (cache *EC2InstanceMetadataCache) getENIsFromPaginatedDescribeNetworkInterfaces(
+	input *ec2.DescribeNetworkInterfacesInput) ([]*ec2.NetworkInterface, error) {
+	outputENIs := make([]*ec2.NetworkInterface, 0)
+	pageNum := 0
+	log.Debugf("Paginating describe ENI has page size: %d", *input.MaxResults)
+	pageFn := func(output *ec2.DescribeNetworkInterfacesOutput, lastPage bool) (nextPage bool) {
+		pageNum++
+		log.Debugf("EC2 DescribeNetworkInterfaces succeeded with %d results on page %d",
+			len(output.NetworkInterfaces), pageNum)
+
+		outputENIs = append(outputENIs, output.NetworkInterfaces...)
```

Comment:
This avoids a huge datastructure on the EC2 server-side (yay), but still builds the same huge list in vpc-cni, just so we can _then_ iterate over the items in the list (boo).  Worse, it's now a function with an API that _can't_ be used in any other way, and it is no longer clear to a caller that this is the way it works internally.

Rather than this proposed helper function, can we call `DescribeNetworkInterfacesPagesWithContext` directly in the caller, and just do the `for _, networkInterface := range result.NetworkInterfaces { ... }` loop body inside the `pageFn` ?

Reply:
> Rather than this proposed helper function, can we call `DescribeNetworkInterfacesPagesWithContext` directly in the caller, and just do the `for _, networkInterface := range result.NetworkInterfaces { ... }` loop body inside the `pageFn` ?

I think recommendation from EC2 is to use MaxResult parameter for DescribeNetworkInterfacesPagesWithContext API?

Reply:
Sorry, my description was probably confusing.  I meant:

```golang
func (cache *EC2InstanceMetadataCache) getENIsFromPaginatedDescribeNetworkInterfaces(
￼	input *ec2.DescribeNetworkInterfacesInput, itemFn func(*ec2.NetworkInterface) bool) error {
    var innerErr error
    pageFn := func(output *ec2.DescribeNetworkInterfacesOutput, lastPage bool) bool {
        for _, eni := range output.NetworkInterfaces {
            if err := itemFn(eni); err != nil {
                innerErr = err
                return false
            }
        }
        return true
    }
    if err := cache.ec2SVC.DescribeNetworkInterfacesPagesWithContext(context.TODO(), input, pageFn, userAgent); err != nil {
        return err
    }
    return innerErr
}
```

... and then back in `getFilteredListOfNetworkInterfaces()`, do:
```golang
        // ...

        var networkInterfaces []*ec2.NetworkInterface
        visitorFn := func(networkInterface *ec2.NetworkInterface) error {
                // [Gus: This is the body of the original `for` loop, with only minor tweaks to control flow]

		// Verify the description starts with "aws-K8S-"
		if !strings.HasPrefix(aws.StringValue(networkInterface.Description), eniDescriptionPrefix) {
			return nil
		}
		// Check that it's not a newly created ENI
		tags := getTags(networkInterface, aws.StringValue(networkInterface.NetworkInterfaceId))
		if value, ok := tags[eniCreatedAtTagKey]; ok {
			parsedTime, err := time.Parse(time.RFC3339, value)
			if err != nil {
				log.Warnf("ParsedTime format %s is wrong so retagging with current TS", parsedTime)
				cache.tagENIcreateTS(aws.StringValue(networkInterface.NetworkInterfaceId), maxENIBackoffDelay)
			}
			if time.Since(parsedTime) < eniDeleteCooldownTime {
				log.Infof("Found an ENI created less than 5 minutes ago, so not cleaning it up")
				return nil
			}
			log.Debugf("%v", value)
		} else {
			/* Set a time if we didn't find one. This is to prevent accidentally deleting ENIs that are in the
			 * process of being attached by CNI versions v1.5.x or earlier.
			 */
			cache.tagENIcreateTS(aws.StringValue(networkInterface.NetworkInterfaceId), maxENIBackoffDelay)
		}

                networkInterfaces = append(networkInterfaces, networkInterface)
		return nil
        }

	input := &ec2.DescribeNetworkInterfacesInput{
￼		Filters: []*ec2.Filter{tagFilter, statusFilter},
￼		MaxResults: aws.Int64(describeENIPageSize),
￼	}
￼	err := getENIsFromPaginatedDescribeNetworkInterfaces(input, visitorFn)
        if err != nil {
￼		return nil, errors.Wrap(err, "awsutils: unable to obtain filtered list of network interfaces")
￼	}

        // .. continue with rest of function, and eventually return networkInterfaces, nil
￼```

```

Reply:
This is so nice and minimizes the size of list holding results data locally. Will try this out. Thanks @anguslees